File: Programmers/Level_2/영어끝말잇기.py
'''
https://programmers.co.kr/learn/courses/30/lessons/12981

문제 명세대로 이미 있는 단어 또는 끝말 잇기가 안 된 단어를 거른다.
해당 단어의 순번을 기준으로 게임 참가인원으로 나머지 연산을 하면 몇 번째 사람이 틀렸는지 알 수 있고
나누기 연산을 통해 몇 번째 라운드인지 알 수 있다.
'''

from math import ceil

def solution(n, words):
    answer = [0, 0]
    lst = []

    for idx, word in enumerate(words):
        if word in lst or (len(lst) > 0 and word[0] != lst[-1][-1]):
            if (idx + 1) % n == 0:
                answer[0] = n
            else:
                answer[0] = (idx + 1) % n
                
            answer[1] = ceil((idx + 1) / n)
            break
        else:
            lst.append(word)

    return answer

File: Programmers/Level_2/Swift_Code/Swift_Code/영어끝말잇기.swift
//
//  영어끝말잇기.swift
//  Swift_Code
//
//  Created by InJe Choi on 2022/04/07.
//

/*
 주어진 조건만 잘 따라서 풀면 쉬운 문제
 몇 번째 사람인지와 몇 번째 차례인지 구하는 것만 살짝 신경쓰면 된다
 */

import Foundation

func solution(_ n: Int, _ words: [String]) -> [Int] {
    var answer = [0, 0]
    var s = Set<String>()
    
    for (idx, word) in words.enumerated() {
        if !s.insert(word).0 || word.count == 1 || (idx > 0 && words[idx-1].last != word.first) {
            let personNumber = (idx + 1) % n == 0 ? n : (idx + 1) % n
            answer = [personNumber, Int(ceil(Double(idx+1) / Double(n)))]
            break
        }
    }
    return answer
}

// print(solution(2, ["hello", "one", "even", "never", "now", "world", "draw"]))
